### ISSUE TITLE
Support new ClassLoaders in JDK 9 and up

### ISSUE THREAD

**Issue by danielcompton** [tobias/dynapath#12]
JDK 9 changed the implementation of ClassLoaders. In particular, the application class loader is no longer an instance of URLClassLoader, but is now an internal class.

```clj
;; Java 8

(->> (iterate #(.getParent ^ClassLoader %) (clojure.lang.RT/baseLoader))
     (take-while identity)
     reverse)
=>
(#object[sun.misc.Launcher$ExtClassLoader 0x5e4cf6a4 "sun.misc.Launcher$ExtClassLoader@5e4cf6a4"]
 #object[sun.misc.Launcher$AppClassLoader 0x18b4aac2 "sun.misc.Launcher$AppClassLoader@18b4aac2"]
 #object[clojure.lang.DynamicClassLoader 0x375dab93 "clojure.lang.DynamicClassLoader@375dab93"]
 #object[clojure.lang.DynamicClassLoader 0x64a2453b "clojure.lang.DynamicClassLoader@64a2453b"]
 #object[clojure.lang.DynamicClassLoader 0x56b1e527 "clojure.lang.DynamicClassLoader@56b1e527"]
 #object[clojure.lang.DynamicClassLoader 0x5a05b614 "clojure.lang.DynamicClassLoader@5a05b614"])

(->> (iterate #(.getParent ^ClassLoader %) (clojure.lang.RT/baseLoader))
     (take-while identity)
     reverse
     (map dc/can-read?))
=> (true true true true true true)


;; Java 11
(->> (iterate #(.getParent ^ClassLoader %) (clojure.lang.RT/baseLoader))
     (take-while identity)
     reverse)
=>
(#object[jdk.internal.loader.ClassLoaders$PlatformClassLoader
         0x7a03e6bc
         "jdk.internal.loader.ClassLoaders$PlatformClassLoader@7a03e6bc"]
 #object[jdk.internal.loader.ClassLoaders$AppClassLoader
         0x512ddf17
         "jdk.internal.loader.ClassLoaders$AppClassLoader@512ddf17"]
 #object[clojure.lang.DynamicClassLoader 0x315f43d5 "clojure.lang.DynamicClassLoader@315f43d5"]
 #object[clojure.lang.DynamicClassLoader 0x75390459 "clojure.lang.DynamicClassLoader@75390459"]
 #object[clojure.lang.DynamicClassLoader 0x6ad3fbe4 "clojure.lang.DynamicClassLoader@6ad3fbe4"]
 #object[clojure.lang.DynamicClassLoader 0x4cff3407 "clojure.lang.DynamicClassLoader@4cff3407"])

(->> (iterate #(.getParent ^ClassLoader %) (clojure.lang.RT/baseLoader))
     (take-while identity)
     reverse
     (map readable-classpath?))
=> (false false true true true true)
```

This was discovered as part of https://github.com/cloverage/cloverage/pull/235.

There is still a [URL Class Path](https://github.com/netroby/jdk9-dev/blob/a2f7bd42b767c299d0e6cba6f3c5d8f4f32f0874/jdk/src/java.base/share/classes/jdk/internal/loader/ClassLoaders.java#L165) internally but I haven't been able to work out how to access it yet.

Adding `--add-opens java.base/jdk.internal.loader=ALL-UNNAMED` to the JVM args will let you get hold of the AppClassLoader:

```clj
(jdk.internal.loader.ClassLoaders/appClassLoader)
```

**Comment by danielcompton** [comment]
This is pretty filthy but it does get you to the URLs on the ClassPath. It requires the `add-opens` flag which means it's not a great method for distribution.

```clj
(defn private-field [obj fn-name-string]
  (let [m (.. obj getClass (getDeclaredField fn-name-string))]
    (. m (setAccessible true))
    (. m (get obj))))

(comment
  (def ucp (private-field (jdk.internal.loader.ClassLoaders/appClassLoader)
                          "ucp"))
  (seq (.getURLs ucp))
  )
```

**Comment by razum2um** [comment]
What do you guys think about replacing `all-classpath-urls` like it does [clojure/java.classpath](https://github.com/clojure/java.classpath/blob/master/src/main/clojure/clojure/java/classpath.clj#L86L87): by just splitting `java.class.path` if it's empty?

**Comment by tobias** [comment]
This has been fixed via #13 and released as 1.1.0.
